Translation of the "Render Props" page (#77)
reactjs/fr.reactjs.org: content/docs/render-props.md

---
id: render-props
title: Props de rendu
permalink: docs/render-props.html
---

> Le terme [« prop de rendu »](https://cdb.reacttraining.com/use-a-render-prop-50de598f11ce) (_render prop_ en anglais, NdT) fait référence à une technique qui consiste à partager du code entre des composants React en utilisant une propriété dont la valeur est une fonction.

Reviewer: ```suggestion
Le terme [« prop de rendu »](https://cdb.reacttraining.com/use-a-render-prop-50de598f11ce) _(render prop, NdT)_ fait référence à une technique qui consiste à partager du code entre des composants React en utilisant une propriété dont la valeur est une fonction.
```

Un composant avec une prop de rendu prend une fonction qui renvoie un élément React et l'appelle au lieu d'implémenter sa propre logique de rendu.

```jsx
<DataProvider render={data => (
  <h1>Bonjour {data.target}</h1>
)}/>
```

Des bibliothèques telles que [React Router](https://reacttraining.com/react-router/web/api/Route/Route-render-methods) et [Downshift](https://github.com/paypal/downshift) utilisent ces props de rendu.

> Dans cet article, nous vous exposerons pourquoi les props de rendu sont pratiques, et comment écrire les vôtres.

Reviewer: ```suggestion
Dans cette page, nous verrons en quoi les props de rendu sont pratiques, et comment écrire les vôtres.
```

On est dans la doc, pas dans un blog / magazine / journal…

## Utiliser les props de rendu pour des questions transversales {#use-render-props-for-cross-cutting-concerns}

Les composants sont l'unité de base de réutilisation du code dans React, mais il n'est pas toujours évident de partager l'état ou le comportement contenu dans un composant avec d'autres composants qui auraient besoin de ce même état.

> Par exemple, le composant suivant traque la position de la souris dans une application web :

Reviewer: ```suggestion
Par exemple, le composant suivant piste la position de la souris dans une application web :
```

« Traque » n'est pas faux, mais ça donne trop une impression de franglais sur ce coup… Par ailleurs, il y a une connotation de chasse, d'un aboutissement qui flingue la cible, ce qui n'est pas le cas de « pister ».

  render() {
    return (
      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>
        <h1>Déplacez votre souris sur l'écran !</h1>

>         <p>La position actuelle de la souris est ({this.state.x}, {this.state.y})</p>

Reviewer: Et m… voilà encore le diff GitHub qui pète un boulon 😒

      </div>
    );
  }
}
```

Lorsque le curseur se déplace sur l'écran, le composant affiche ses coordonnées (x,y) dans un `<p>`.

> La question qui se pose maintenant est : Comment pouvons-nous réutiliser ce comportement dans un autre composant ? En d'autres termes, si un autre composant a besoin de connaître la position du curseur, pouvons-nous encapsuler ce comportement pour pouvoir facilement le partager avec ce composant ?

Reviewer: ```suggestion
La question qui se pose maintenant est : comment pouvons-nous réutiliser ce comportement dans un autre composant ? En d'autres termes, si un autre composant a besoin de connaître la position du curseur, pouvons-nous encapsuler ce comportement pour pouvoir facilement le partager avec ce composant ?
```

Pas de majuscules après un deux-points en français.

    return (
      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>

>         {/* ...mais comment rendre autre chose qu'un <p> ? */}

Reviewer: ```suggestion
        {/* ...mais comment afficher autre chose qu'un <p> ? */}
```

  render() {
    return (
      <div>
        <h1>Déplacez votre souris sur l'écran !</h1>
        <Mouse />
      </div>
    );
  }
}
```

> Le composant `<Mouse>` encapsule maintenant tous les comportements associés à l'écoute des évènements `mousemove` et à la sauvegarde de la position (x, y) du sauvegarde, mais il n'est pas encore tout à fait réutilisable.

Reviewer: ```suggestion
Le composant `<Mouse>` encapsule maintenant tous les comportements associés à l'écoute des événements `mousemove` et à la sauvegarde de la position (x, y) du curseur, mais il n'est pas encore tout à fait réutilisable.
```

On conserve l'orthographe traditionnelle (et toujours valide) de « événement »

> Par exemple, supposons que nous avons un composant `<Cat>` qui rend une image de chat chassant une souris sur l'écran. Nous pourrions utiliser une propriété `<Cat mouse={{ x, y }}>` pour transmettre au composant les coordonnées de la souris pour qu'il sache où positionner l'image sur l'écran.

Reviewer: ```suggestion
Par exemple, supposons que nous avons un composant `<Cat>` qui affiche une image de chat pourchassant une souris sur l'écran. Nous pourrions utiliser une prop `<Cat mouse={{ x, y }}>` pour transmettre au composant les coordonnées de la souris pour qu'il sache où positionner l'image sur l'écran.
```

> Au premier essai, vous tenteriez peut-être de rendre `<Cat>` *dans la méthode `render` de `<Mouse>`*, comme ceci :

Reviewer: ```suggestion
Au premier essai, vous tenteriez peut-être d’afficher `<Cat>` *dans la méthode `render` de `<Mouse>`*, comme ceci :
```

```js
class Cat extends React.Component {
  render() {
    const mouse = this.props.mouse;
    return (

>       <img src="/chat.jpg" style={{ position: 'absolute', left: mouse.x, top: mouse.y }} />

Reviewer: ```suggestion
      <img src="/cat.jpg" style={{ position: 'absolute', left: mouse.x, top: mouse.y }} />
```

Les noms de fichiers sont comme les identifiants du code : on ne traduit pas.

  render() {
    return (
      <div>
        <h1>Déplacez votre souris sur l'écran !</h1>
        <MouseWithCat />
      </div>
    );
  }
}
```

> Cette approche fonctionnera dans notre cas particulier, mais nous n'avons pas atteint notre objectif qui consiste à vraiment encapsuler le comportement de façon réutilisable. En effet, chaque fois que nous aurons besoin de la position de la souris pour un cas d'utilisation différent, nous devrons créer un nouveau composant (i.e. pour ainsi dire un autre `<MouseWithCat>`) spécifique à ce cas.

Reviewer: ```suggestion
Cette approche fonctionnera dans notre cas particulier, mais nous n'avons pas atteint notre objectif qui consiste à vraiment encapsuler le comportement de façon réutilisable. En effet, chaque fois que nous aurons besoin de la position de la souris pour un cas d'utilisation différent, nous devrons créer un nouveau composant (pour ainsi dire un autre `<MouseWithCat>`) spécifique à ce cas.
```

`i.e.` n'est pas acceptable en français ; on peut le remplacer le plus souvent par « c'est-à-dire » (on évitera l'abréviation « c.-à-d. »), mais parfois, comme ici, on peut juste le virer sans y perdre.

> Cette approche fonctionnera dans notre cas d'utilisation, mais nous n'avons pas atteint notre objectif de vraiment encapsuler le comportement de façon réutilisable. En effet, chaque fois que nous aurons besoin de la position de la souris pour un cas d'utilisation différent, nous devrons créer un nouveau composant (i.e. pour ainsi dire un autre `<MouseWithCat>`) that renders something specifically for that use case.

Reviewer 2: ```suggestion
Cette approche fonctionnera dans notre cas particulier, mais nous n'avons pas atteint notre objectif qui consiste à vraiment encapsuler le comportement de façon réutilisable. En effet, chaque fois que nous aurons besoin de la position de la souris pour un cas d'utilisation différent, nous devrons créer un nouveau composant (i.e. pour ainsi dire un autre `<MouseWithCat>`) spécifique à ce cas.
```

> C'est là que la prop de rendu entre en scène : au lieu d'écrire en dur un composant `<Cat>` dans un composant `<Mouse>`, et changer la valeur de sortie rendue, nous pouvons créer `<Mouse>` avec une propriété qui prendra une fonction qui permet de déterminer dynamiquement ce qui doit être rendu—une prop de rendu.

Reviewer: ```suggestion
C'est là que la prop de rendu entre en scène : au lieu d'écrire en dur un composant `<Cat>` dans un composant `<Mouse>`, et changer le résultat de l’affichage, nous pouvons créer `<Mouse>` avec une prop qui prendra une fonction permettant de déterminer dynamiquement ce qui doit être affiché : une prop de rendu.
```

> C'est là que la propriété render entre en scène : au lieu d'écrire en dur un composant `<Cat>` dans un composant `<Mouse>`, et changer la valeur de sortie rendue, nous pouvons fournir `<Mouse>` avec une propriété qui est une fonction qui permet de déterminer dynamiquement ce qui doit être rendu, c'est-à-dire une render prop.

Reviewer 2: ```suggestion
C'est là que la prop de rendu entre en scène : au lieu d'écrire en dur un composant `<Cat>` dans un composant `<Mouse>`, et changer la valeur de sortie rendue, nous pouvons créer `<Mouse>` avec une propriété qui prendra une fonction qui permet de déterminer dynamiquement ce qui doit être rendu—une prop de rendu.
```

```js
class Cat extends React.Component {
  render() {
    const mouse = this.props.mouse;
    return (

>       <img src="/chat.jpg" style={{ position: 'absolute', left: mouse.x, top: mouse.y }} />

Reviewer: ```suggestion
      <img src="/cat.jpg" style={{ position: 'absolute', left: mouse.x, top: mouse.y }} />
```

      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>

        {/*

>           Au lieu de fournir un représentation statique de ce que rend <Mouse>,

Reviewer: ```suggestion
          Au lieu de fournir un représentation statique de ce qu’affiche <Mouse>,
```

>           utilisez la prop `render` pour déterminer dynamiquement quoi rendre.

Reviewer: ```suggestion
          utilisez la prop `render` pour déterminer dynamiquement quoi afficher.
```

}
```

> Dorénavant, au lieu de cloner le composant `<Mouse>` et d'écrire en dur quelque chose d'autre dans sa méthode `render` pour s'adapter à un cas d'utilisation, nous fournissons une prop `render` que `<Mouse>` peut utiliser pour déterminer dynamiquement quoi rendre.

Reviewer: ```suggestion
Dorénavant, au lieu de cloner le composant `<Mouse>` et d'écrire en dur quelque chose d'autre dans sa méthode `render` pour s'adapter à un cas d'utilisation, nous fournissons une prop `render` que `<Mouse>` peut utiliser pour déterminer dynamiquement quoi afficher.
```

> Plus concrètement, **une prop de rendu est une propriété fonction qu'un composant utilise pour savoir quoi rendre.**

Reviewer: ```suggestion
Plus concrètement, **une prop de rendu est une prop de type fonction qu'un composant utilise pour savoir quoi afficher.**
```

> Cette technique facilite énormément le partage d'un comportement. Pour récupérer ce comportement, il suffit de faire le rendu d'un `<Mouse>` avec une propriété `render` qui lui dit quoi rendre avec les coordonnées (x, y) actuelles du curseur.

Reviewer: ```suggestion
Cette technique facilite énormément le partage d'un comportement. Pour récupérer ce comportement, il suffit d’afficher un `<Mouse>` avec une prop `render` qui lui dit quoi afficher avec les coordonnées (x, y) actuelles du curseur.
```

> Un point intéressant à noter concernant les props de rendu est que vous pouvez implémenter la plupart des [composants d’ordre supérieur](/docs/higher-order-components.html) (HOC) en utilisant un composant classique avec une prop de rendu. Par exemple, si vous préférez avoir un HOC `withMouse` au lieu d'un composant`<Mouse>`, vous pouvez facilement en créer un en utilisant un composant `<Mouse>` avec une prop de rendu :

Reviewer: ```suggestion
Un point intéressant à noter concernant les props de rendu est que vous pouvez implémenter la plupart des [composants d’ordre supérieur](/docs/higher-order-components.html) *(Higher-Order Components, ou HOC, NdT)* en utilisant un composant classique avec une prop de rendu. Par exemple, si vous préférez avoir un HOC `withMouse` au lieu d'un composant`<Mouse>`, vous pouvez facilement en créer un en utilisant un composant `<Mouse>` avec une prop de rendu :
```

}
```

Utiliser une prop de rendu rend donc possible l'utilisation des deux méthodes.

> ## Utiliser d'autres propriétés que `render` {#using-props-other-than-render}

Reviewer: ```suggestion
## Utiliser d'autres props que `render` {#using-props-other-than-render}
```

> Il est important de se rappeler que *ce n'est pas parce que la méthode s'appelle "props de rendu" qu'il est obligatoire d'utiliser une propriété appelée `render` pour utiliser cette méthode*. En fait, [*n'importe quelle* propriété qui est utilisée par un composant pour savoir quoi renvoyer est techniquement une "prop de rendu" ](https://cdb.reacttraining.com/use-a-render-prop-50de598f11ce).

Reviewer 2: ```suggestion
Il est important de se rappeler que *ce n'est pas parce que la méthode s'appelle « props de rendu » qu'il est obligatoire d'utiliser une propriété appelée `render` pour utiliser cette méthode*. En fait, [*n'importe quelle* propriété qui est utilisée par un composant pour savoir quoi renvoyer est techniquement une « prop de rendu » ](https://cdb.reacttraining.com/use-a-render-prop-50de598f11ce).
```

> Il est important de se rappeler que *ce n'est pas parce que la méthode s'appelle « props de rendu » qu'il est obligatoire d'utiliser une propriété appelée `render` pour utiliser cette méthode*. En fait, [*n'importe quelle* propriété qui est utilisée par un composant pour savoir quoi renvoyer est techniquement une « prop de rendu » ](https://cdb.reacttraining.com/use-a-render-prop-50de598f11ce).

Reviewer: ```suggestion
Il est important de se rappeler que *ce n'est pas parce que la technique s'appelle « props de rendu » qu'il est obligatoire d'utiliser une prop appelée `render` pour la mettre en œuvre*. En fait, [*n'importe quelle* prop utilisée par un composant pour savoir quoi renvoyer est techniquement une « prop de rendu »](https://cdb.reacttraining.com/use-a-render-prop-50de598f11ce).
```

> Même si les exemples ci-dessus utilise `render`, nous pourrions tout aussi simplement utiliser la propriété `children` !

Reviewer: ```suggestion
Même si les exemples ci-dessus utilisent `render`, nous pourrions tout aussi simplement utiliser la prop `children` !
```

```js
<Mouse children={mouse => (
  <p>La position de la souris est {mouse.x}, {mouse.y}</p>
)}/>
```

> Et rappelez-vous, la propriété `children` n'a en fait pas besoin d'être nommée dans la liste des "attributs" de votre élément JSX. Au lieu de ça, vous pouvez l'utiliser directement *dans* l'élément !

Reviewer: ```suggestion
Et rappelez-vous, la propriété `children` n'a en fait pas besoin d'être nommée dans la liste des « attributs » de votre élément JSX. Au lieu de ça, vous pouvez l'utiliser directement *dans* l'élément !
```

</Mouse>

> ```

Reviewer: Ahem, pense à reprendre la traduction dans le `<p>`ci-dessus…

Author: 🤦🏻‍♀️ je corrige ça !

> Vous pouvez découvrir l'utilisation de cette méthode dans l'API [react-motion](https://github.com/chenglou/react-motion).

Reviewer: ```suggestion
Vous pouvez découvrir l'utilisation de cette technique dans l'API [react-motion](https://github.com/chenglou/react-motion).
```

Faut arrêter avec « méthode » pour « technique », alors que « méthode » a une connotation bien particulière dans du code objet…

> Comme cette méthode est un peu inhabituelle, vous aurez probablement envie de préciser que `children` devrait être une fonction dans vos `propTypes` au moment de faire le design de votre API de la manière suivante.

Reviewer: ```suggestion
Comme cette technique est un peu inhabituelle, vous aurez probablement envie de préciser que `children` devrait être une fonction dans vos `propTypes` au moment de concevoir votre API de cette façon.
```

```js
Mouse.propTypes = {
  children: PropTypes.func.isRequired
};
```

> ## Exceptions {#caveats}

Reviewer: ```suggestion
## Limitations {#caveats}
```

### Soyez prudents lors de l'utilisation de props de rendu avec React.PureComponent {#be-careful-when-using-render-props-with-reactpurecomponent}

> Utiliser une prop de rendu peut contrebalancer l'avantage apporté par l'utilisation de [`React.PureComponent`](/docs/react-api.html#reactpurecomponent) si vous créer une fonction dans une méthode `render`. Cela est dû au fait que la comparaison superficielle des propriétés retournera toujours `false` pour les nouvelles propriétés, et que dans ce cas, chaque `render` génèrera une nouvelle valeur pour la prop de rendu.

Reviewer: ```suggestion
Utiliser une prop de rendu peut contrebalancer l'avantage apporté par l'utilisation de [`React.PureComponent`](/docs/react-api.html#reactpurecomponent) si vous créez la fonction dans une méthode `render`. C’est dû au fait que la comparaison superficielle des propriétés renverra toujours `false` pour les nouvelles props, et que dans ce cas chaque `render` génèrera une nouvelle valeur pour la prop de rendu.
```

Par exemple, en gardant l'exemple de notre composant `<Mouse>`, si `Mouse` étendait `React.PureComponent` au lieu de `React.Component`, notre exemple ressemblerait alors à ça :

```js
class Mouse extends React.PureComponent {
  // Même implémentation que plus haut...
}

class MouseTracker extends React.Component {
  render() {
    return (
      <div>
        <h1>Déplacez votre souris sur l'écran !</h1>

        {/*

>           Il ne faut pas faire ça ! La valeur de la propriété `render`

Reviewer: ```suggestion
          Il ne faut pas faire ça ! La valeur de la prop `render`
```

**Parfait** 💯  « il ne faut pas faire ça »

>           sera différente à chaque render.

Reviewer: ```suggestion
          sera différente à chaque rendu.
```

}
```

> Dans cet exemple, à chaque fois que `<MouseTracker>` est rendu, une nouvelle fonction est générée comme valeur de la propriété `<Mouse render>`, ce qui annule l'effet de `<Mouse>` qui étend `React.PureComponent` !

Reviewer: ```suggestion
Dans cet exemple, à chaque fois que `<MouseTracker>` s’affiche, une nouvelle fonction est générée comme valeur de la prop `<Mouse render>`, ce qui annule l'optimisation recherchée à la base quand on a fait en sorte que `<Mouse>` étende `React.PureComponent` !
```

> Pour éviter ce problème; vous pouvez parfois définir la propriété comme une méthode d'instance, de la façon suivante :

Reviewer: ```suggestion
Pour éviter ce problème; vous pouvez parfois définir la prop comme une méthode d'instance, de la façon suivante :
```

```js
class MouseTracker extends React.Component {
  // Définie comme une méthode d'instance, `this.renderTheCat` se réfèrera
  // toujours à la *même* fonction quand nous l'utiliserons dans le rendu
  renderTheCat(mouse) {
    return <Cat mouse={mouse} />;
  }

  render() {
    return (
      <div>
        <h1>Déplacez votre souris sur l'écran !</h1>
        <Mouse render={this.renderTheCat} />
      </div>
    );
  }
}
```

> Dans les cas où vous ne pourrez pas définir la propriété de manière statique (e.g. parce que vous devez fermer la propriété et/ou l'état local) `<Mouse>` devrait plutôt étendre `React.Component`.

Reviewer: ```suggestion
Dans les cas où vous ne pourrez pas définir la prop de manière statique (ex. parce qu’elle utilise des valeurs locales à `render`), `<Mouse>` devrait plutôt étendre `React.Component`.
```

Les raisons invoquées sont en fait difficiles à capter, puisqu'on est dans une classe et non une fonction composant : les props et l'état local sont donc accessibles via `this`, même si on y perd la garantie de l'identité du rendu. C'est plus pratique de parler des identifiants locaux du `render` pour illustrer la question.
